### PR TITLE
Autosave option for alert filters #1

### PR DESCRIPTION
**Pull request by dmitryduev**
This PR implements the option to automatically save passing alerts/candidates to corresponding groups as Sources on SP in broker mode.

Will implement rate limiting in the next PR in this series. (see the discussion below and https://github.com/dmitryduev/kowalski/issues/54)

### CODE REVIEW

**Comment by profjsb** [comment]
Wondering if we should have a global max per time unit for this functionality that forbids saving more than this per filter. Very worried about a permissive filter polluting the sources in SP. Say, no more than 5000 sources per 24 hr per filter?

SP might want to implement something like this as well, per API token.

File: kowalski/alert_broker_ztf.py [hunk 1]
```diff
@@ -901,6 +903,10 @@ def api_skyportal(self, method: str, endpoint: str, data=None):
 
         return response
 
+    @lru_cache(maxsize=16)
```
Comment (stefanv) [inline kowalski/alert_broker_ztf.py:906]: It's probably cheap to make this a bit larger?

File: kowalski/api.py [hunk 2]
```diff
@@ -2715,22 +2744,33 @@ async def filters_put(request):
...
-                # note: this would fire up if, e.g. trying to change from True to True
-                return web.json_response(
-                    {"status": "error", "message": "failed to update filter"},
-                    status=400,
-                )
+            return_data = {"active_fid": active_fid}
+
+        elif autosave is not None:
+            # be forgiving:
+            autosave = True if autosave in ("t", "True", "true", "1", True) else False
```
Comment (stefanv) [inline kowalski/api.py:2751]: Heh, this should probably be a utility function.

**Review by stefanv — COMMENTED** [review]
This looks good to me, @dmitryduev.  Only question, from a design perspective, is whether this introduces a security hole.

**Comment by dmitryduev** [comment]
@profjsb I agree, that would be a good measure, I'll think about the implementation. Do you think it should be part of this PR?

@stefanv could you please elaborate on what you think might pose a security threat?

**Comment by profjsb** [comment]
Maybe make this PR number one of two, and say in the comments that you intend to add rate limiting in the sibling PR

File: kowalski/alert_broker_ztf.py (hunk 1 above)
Comment (dmitryduev) [reply kowalski/alert_broker_ztf.py:906]: I'll bump `maxsize` a bit, but it's actually done on purpose -- it is still a good idea to check from time to time, just not every time.

**Comment by stefanv** [comment]
I just wanted to check that we've thought through how these filters will come into place, that users without permissions to post to a group cannot set up filters that post to other groups (presumably filter posting will happen from SkyPortal with a token, so should be ok).

**Comment by dmitryduev** [comment]
Ah, yea, that's always a good idea @stefanv :) The endpoint on the K side requires the user/token to be admin, which regular K users are not. On the F side (and F's token is that of an admin from the K point of view), the interaction with K is handled [here](https://github.com/fritz-marshal/fritz/blob/master/extensions/skyportal/skyportal/handlers/api/kowalski_filter.py) and all the permissions checks are supposedly in place (e.g. you must be the group's admin in order to save/modify the filter).

**Review by kmshin1397 — APPROVED** [review]
LGTM! Good to see the Kowalski tests being expanded :)

Agree with Josh that a follow-up PR to add rate limiting is needed.

**Comment by dmitryduev** [comment]
Got a verbal approval from @stefanv, so self-merging.

**Pull request merged**